Docs: Add `sveltekit` framework doc (#26329)
storybookjs/storybook: docs/get-started/sveltekit.md

            console.log(to, event);
          },
          asRegex: true,
        },
      },
    },
  },
};
```

See the [API reference](#hrefs) for more information.

## Writing native Svelte stories

Storybook provides a Svelte addon maintained by the community, enabling you to write stories for your Svelte components using the template syntax. You'll need to take some additional steps to enable this feature.

Run the following command to install the addon.

<!-- prettier-ignore-start -->

<CodeSnippets
  paths={[
   'svelte/svelte-csf-addon-install.yarn.js.mdx',
   'svelte/svelte-csf-addon-install.npm.js.mdx',
   'svelte/svelte-csf-addon-install.pnpm.js.mdx',
  ]}
/>

<!-- prettier-ignore-end -->

<Callout variant="info">

The community actively maintains the Svelte CSF addon but still lacks some features currently available in the official Storybook Svelte framework support. For more information, see [the addon's documentation](https://github.com/storybookjs/addon-svelte-csf).

</Callout>

## API

### Parameters

This framework contributes the following [parameters](../writing-stories/parameters.md) to Storybook, under the `sveltekit_experimental` namespace:

#### `forms`

> Type: `{ enhance: TK }`

Reviewer: ```suggestion
Type: `{ enhance: () => void }`
```

Provides mocks for the `$app/forms` module.

##### `forms.enhance`

> Type: TK

Reviewer: ```suggestion
Type: () => void
```

A callback that will be called when a form with `use:enhance` is submitted.

#### `hrefs`

> Type: `Record<[path: string], (to: string, event: TK) => void | { callback: (to: string, event: TK) => void, asRegex?: boolean }>`

Reviewer: ```suggestion
Type: `Record<[path: string], (to: string, event: MouseEvent) => void | { callback: (to: string, event: MouseEvent) => void, asRegex?: boolean }>`
```

If you have an `<a />` tag inside your code with the `href` attribute that matches one or more of the links defined (treated as regex based if the `asRegex` property is `true`) the corresponding `callback` will be called. If no matching `hrefs` are defined, an action will be logged to the [Actions panel](../essentials/actions.md). See [Mocking links](#mocking-links) for an example.

#### `navigation`

Type: TK

Provides mocks for the `$app/navigation` module.

##### `navigation.goto`

> Type: TK

Reviewer: This would be the same as SvelteKit's own `goto` type - I don't know how to document that best? 

https://github.com/sveltejs/kit/blob/main/packages/kit/types/index.d.ts#L2055-L2068

A callback that will be called whenever `goto` is called. If no function is provided, an action will be logged to the [Actions panel](../essentials/actions.md).

##### `navigation.pushState`

> Type: TK

Reviewer: https://github.com/sveltejs/kit/blob/main/packages/kit/types/index.d.ts#L2121-L2125

A callback that will be called whenever `pushState` is called. If no function is provided, an action will be logged to the [Actions panel](../essentials/actions.md).

##### `navigation.replaceState`

> Type: TK

Reviewer: https://github.com/sveltejs/kit/blob/main/packages/kit/types/index.d.ts#L2126-L2130

A callback that will be called whenever `replaceState` is called. If no function is provided, an action will be logged to the [Actions panel](../essentials/actions.md).

##### `navigation.invalidate`

> Type: TK

Reviewer: https://github.com/sveltejs/kit/blob/main/packages/kit/types/index.d.ts#L2087-L2090

A callback that will be called whenever `invalidate` is called. If no function is provided, an action will be logged to the [Actions panel](../essentials/actions.md).

##### `navigation.invalidateAll`

> Type: TK

Reviewer: https://github.com/sveltejs/kit/blob/main/packages/kit/types/index.d.ts#L2087-L2090

A callback that will be called whenever `invalidateAll` is called. If no function is provided, an action will be logged to the [Actions panel](../essentials/actions.md).

##### `navigation.afterNavigate`

> Type: TK

Reviewer: https://github.com/sveltejs/kit/blob/main/packages/kit/types/index.d.ts#L2022-L2027

An object that will be passed to the `afterNavigate` function, which will be invoked when the `onMount` event fires.

> #### `stores`

Reviewer: Similar to above, maybe linking to SvelteKit's docs on the types is best?

https://kit.svelte.dev/docs/modules#$app-stores

---
title: Storybook for SvelteKit
---

export const SUPPORTED_RENDERER = 'svelte';

Storybook for SvelteKit is a [framework](../contribute/framework.md) that makes it easy to develop and test UI components in isolation for [SvelteKit](https://kit.svelte.dev/) applications. It includes:

- 🪄 Zero config
- 🧩 Easily mock many Kit modules
- 🔗 Automatic link handling
- 💫 and more!

<If notRenderer={SUPPORTED_RENDERER}>

<Callout variant="info">

Storybook for SvelteKit is only supported in [Svelte](?renderer=svelte) projects.

</Callout>

<!-- End non-supported renderers -->

</If>

<If renderer={SUPPORTED_RENDERER}>

## Requirements

> - SvelteKit ≥ 1.0 (not including beta versions)

Reviewer: I don't think we need to state the part about beta versions anymore, they are long gone.

```suggestion
- SvelteKit ≥ 1.0
```

| [`@sveltejs/kit/*`](https://kit.svelte.dev/docs/modules#sveltejs-kit)              | ✅ Supported           |                                                                                                                                         |
| [`$env/dynamic/private`](https://kit.svelte.dev/docs/modules#$env-dynamic-private) | ⛔ Not supported       | This is a server-side feature, and Storybook renders all components on the client.                                                      |
| [`$env/static/private`](https://kit.svelte.dev/docs/modules#$env-static-private)   | ⛔ Not supported       | This is a server-side feature, and Storybook renders all components on the client.                                                      |
| [`$service-worker`](https://kit.svelte.dev/docs/modules#$service-worker)           | ⛔ Not supported       | This is a service worker feature, which does not apply to Storybook.                                                                    |

> ## How to mock

Reviewer: perhaps add your usual "⚠️ Experimental" jazz here.

Author: @JReinhold — Hmm... I'm having a hard time thinking through where this best fits (besides "everywhere").

What would you think about replacing "✅ Supported" with "⚠️ Experimental" in the table above, for relevant features? If you like that approach, I need your help knowing which are which, please.

Reviewer: That sounds good to me. It would be lines 117, 118 and 120

To mock a SvelteKit import you can define it within `parameters.sveltekit_experimental`:

```ts
// MyComponent.stories.svelte
export const MyStory = {
  parameters: {
    sveltekit_experimental: {
      stores: {
        page: {
          data: {
            test: 'passed',
          },
        },
        navigating: {
          route: {
            id: '/storybook',
          },
        },
        updated: true,
      },
    },
  },
};
```

The [available parameters](#parameters) are documented in the API section, below.

### Mocking links

The default link-handling behavior (e.g. when clicking an `<a href="..." />` element) is to log an action to the [Actions panel](../essentials/actions.md).

> You can override this by assigning an object to `parameters.sveltekit_experimental.hrefs`, where the keys are strings representing an href and the values define your mock`. For example:

Reviewer: Would it make sense somewhere in this paragraph to link to the `hrefs` API docs just below?

```ts
// MyComponent.stories.svelte
export const MyStory = {
  parameters: {
    sveltekit_experimental: {
      hrefs: {
        '/basic-href': (to, event) => {
          console.log(to, event);
        },
        '/root.*': {
          callback: (to, event) => {
            console.log(to, event);
          },
          asRegex: true,
        },
      },
    },
  },
};
```

## Writing native Svelte stories

Storybook provides a Svelte addon maintained by the community, enabling you to write stories for your Svelte components using the template syntax. You'll need to take some additional steps to enable this feature.

Run the following command to install the addon.

<!-- prettier-ignore-start -->

<CodeSnippets
  paths={[
   'svelte/svelte-csf-addon-install.yarn.js.mdx',
   'svelte/svelte-csf-addon-install.npm.js.mdx',
   'svelte/svelte-csf-addon-install.pnpm.js.mdx',
  ]}
/>

<!-- prettier-ignore-end -->

<Callout variant="info">

> The community actively maintains the Svelte CSF addon but still lacks some features currently available in the official Storybook Svelte framework support. For more information, see [addon's documentation](https://github.com/storybookjs/addon-svelte-csf).

Reviewer: ```suggestion
The community actively maintains the Svelte CSF addon but still lacks some features currently available in the official Storybook Svelte framework support. For more information, see [the addon's documentation](https://github.com/storybookjs/addon-svelte-csf).
```

</Callout>

## API

### Parameters

This framework contributes the following [parameters](../writing-stories/parameters.md) to Storybook, under the `sveltekit_experimental` namespace:

#### `forms`

Type: `{ enhance: TK }`

Provides mocks for the `$app/forms` module.

##### `forms.enhance`

> Type: TK

Reviewer: What is type TK in this whole section?

Author: Thanks for all the suggestions! I opted to handle them in a single commit along with some other updates, so it appears I did not accept them. But I did!

> A callback that will called when a form with `use:enhance` is submitted.

Reviewer: ```suggestion
A callback that will be called when a form with `use:enhance` is submitted.
```

#### `hrefs`

Type: `Record<[path: string], (to: string, event: TK) => void | { callback: (to: string, event: TK) => void, asRegex?: boolean }>`

> If you have an `<a />` tag inside your code with the `href` attribute that matches one or more of the links defined (treated as regex based on the `asRegex` property) the corresponding `callback` will be called. See [Mocking links](#mocking-links) for an example.

Reviewer: ```suggestion
If you have an `<a />` tag inside your code with the `href` attribute that matches one or more of the links defined (treated as regex based if the `asRegex` property is `true`) the corresponding `callback` will be called. If no matching `hrefs` are defined, an action will be logged to the [Actions panel](../essentials/actions.md). See [Mocking links](#mocking-links) for an example.
```

That last "If no matching...." part is just for consistency with `navigation.goto` below.

##### `navigation.replaceState`

Type: TK

A callback that will be called whenever `replaceState` is called. If no function is provided, an action will be logged to the [Actions panel](../essentials/actions.md).

##### `navigation.invalidate`

Type: TK

A callback that will be called whenever `invalidate` is called. If no function is provided, an action will be logged to the [Actions panel](../essentials/actions.md).

##### `navigation.invalidateAll`

Type: TK

A callback that will be called whenever `invalidateAll` is called. If no function is provided, an action will be logged to the [Actions panel](../essentials/actions.md).

##### `navigation.afterNavigate`

Type: TK

An object that will be passed to the `afterNavigate` function, which will be invoked when the `onMount` event fires.

#### `stores`

Type: `{ navigating?: {}, page?: {}, updated?: boolean }`

Provides mocks for the `$app/stores` module.

##### `stores.navigating`

Type: TK

A partial version of the `navigating` store.

##### `stores.page`

Type: TK

A partial version of the `page` store.

##### `stores.updated`

Type: boolean

> A boolean representing the value of `updated` (you can also access `stores.check()` which will be a noop).

Reviewer: ```suggestion
A boolean representing the value of `updated` (you can also access `stores.check()` which will be a no-op).
```

### Options

You can pass an options object for additional configuration if needed:

```js
// .storybook/main.js
import * as path from 'path';

export default {
  // ...
  framework: {
    name: '@storybook/sveltekit',
    options: {
      // ...
    },
  },
};
```

The available options are:

#### `builder`

Type: `Record<string, any>`

Configure options for the [framework's builder](../api/main-config-framework.md#optionsbuilder). For Sveltekit, available options can be found in the [Vite builder docs](../builders/vite.md).

## Troubleshooting

### Error when starting Storybook

When starting Storybook after upgrading to v7.0, it may quit with the following error:

```sh
ERR! SyntaxError: Identifier '__esbuild_register_import_meta_url__' has already been declared
```

This can occur when manually upgrading from 6.5 to 7.0. To resolve it, you'll need to remove the `svelteOptions` property in `.storybook/main.js`, as that is not supported (and no longer necessary) in Storybook 7+ with SvelteKit.

> ### Error in preview canvas
> 
> When running Storybook, some stories may not load. Instead they show the following error in the preview canvas:
> 
> ```sh
> Cannot read properties of undefined (reading 'disable_scroll_handling')
> ```
> 
> You'll experience this if anything in your story is importing from `$app/forms` or `$app/navigation`, which is currently not supported. To get around this, separate your component into a shallow parent component that imports what's needed and passes it to a child component via props. This way you can write stories for your child component and mock any of the necessary modules by passing props in.
> 

Reviewer: This error is no longer relevant since https://github.com/storybookjs/storybook/pull/25138, 8.0.0-alpha.2 . Was there a deliberate reason to include this?

```suggestion
```
